connectors/llm/interface.py

```diff
@@ -55,30 +64,61 @@ def ask(self, system_prompt, previous_messages, question, agent_id, stream):
                     msg_list.append(AIMessage(content=f"{msg['text']}</s>"))
         prompt.messages = msg_list + prompt.messages
 
-        log.debug("prompt: %s", prompt)
-        model = ChatOpenAI(
+        llm = ChatOpenAI(
             model=cfg.LLM_MODEL_NAME,
             openai_api_base=cfg.LLM_BASE_URL,
             openai_api_key=cfg.LLM_API_KEY,
             temperature=cfg.LLM_TEMPERATURE,
         )
 
-        chain = prompt | model | StrOutputParser()
+        chain = prompt | llm
+
+        log.debug("prompting llm...")
+
+        def get_llm_response():
+            timer_start = None
+
+            with get_openai_callback() as cb:
+                for chunk in chain.stream(prompt_params, stream_usage=True):
+                    if not timer_start:
+                        # this is the first output token received
+                        timer_start = time.time()
+                    if len(chunk.content):
+                        text_content = {"text_content": chunk.content}
+                        yield text_content
+
+                timer_end = time.time()
+                search_metadata = {"search_metadata": extra_doc_info}
+                yield search_metadata
+
+            output_rate = cb.completion_tokens / (timer_end - timer_start)
```

Comment:
I *think* there's the possibility of divide by zero here in a couple different ways, if completion_tokens is 0 or timer_end - timer_start is 0.